### ISSUE TITLE
"Forbidden" error when adding or refreshing APKMirror track-only apps

### ISSUE THREAD

**Issue by bossanovaorca** [ImranR98/Obtainium#1973]
**Prerequisites**

**Describe the bug**

APKMirror applications are track-only on Obtainium, but when refreshing or trying to install new apps on APKMirror, Obtainium returns "Forbidden". It doesn't seem to be related to my network so I'm not sure why Obtainium is having trouble with this site.

**To Reproduce**

1. Add an APKMirror application to Obtainium, for example https://www.apkmirror.com/apk/x-corp/twitter/ (it will be "track-only")
2. look for any "Forbidden" errors

**Please complete the following information:**
 - Device: Samsung Galaxy S23 Ultra
 - OS: Android 14
 - Obtainium Version: 1.1.30

I can provide any logs or screenshots if needed

**Comment by leo-liar** [comment]
Can report the same behaviour since a few days ago (with a different app). Update check suddenly failed with "Forbidden". Removed the app from Obtainium but could not add it again because this action is also "Forbidden".

- Device: OnePlus 8T
- OS: Android 13
- Obtainium version: 1.1.30

**Comment by ImranR98** [comment]
Yeah it seems like cloudflare protection has gone from being inconsistent to always on. May have to remove the apkmirror source if a solution is not found.

**Comment by bossanovaorca** [comment]
oh ok I did see #1097 when searching for previous tickets but I wasn't aware that my current issue was Cloudflare related. I may have to look for other sources to migrate to then

**Comment by DwainZwerg** [comment]
@archon810 Do you have an idea?

**Comment by DwainZwerg** [comment]
Do APK Grabber and APK Updater also have the problem? Wenn nein, könnten wir deren source Code übernehmen.

**Comment by archon810** [comment]
Try setting a proper user agent that references the app being used to query ("Obtainium" with the version would be a good start).

**Comment by fahim-ahmed05** [comment]
Try with common browser user agent

**Comment by Vinfall** [comment]
Pretty much a *intended behavior* if I have to guess.

In the past APKMirror also blocked apkeep by EFF, see `https://github.com/EFForg/apkeep/issues/46#issuecomment-1308244059` (quoted to avoid linking issue). But that was 2022, things might have changed.

Also, since APKMirror source in Obtainium is track-only, users need to visit APKMirror site for downloading, maybe they would (still) like to

> welcome and collaborate with apps that try to integrate with the site and direct users to our download pages, or even facilitate an API level collaboration

**Comment by aleks01100001** [comment]
> Try with common browser user agent 

Can we change the user agent ourselves?

**Comment by DwainZwerg** [comment]
Yes, in the settings of the respective app.

**Comment by leo-liar** [comment]
> Yes, in the settings of the respective app.

When adding a non-HTML-source app (like one from APKMirror) this is _not_ possible.

**Comment by fahim-ahmed05** [comment]
> Yes, in the settings of the respective app.

Don't see any option to change the user agent

**Comment by ImranR98** [comment]
Latest version now correctly identifies Obtainium in the user agent (with no option to override). @archon810 please let us know if there's anything else that can be done. If not the source should be removed.

**Comment by archon810** [comment]
@ImranR98 Now that there's a proper UA, I took a look at the Cloudflare WAF stats. I found that the vast majority of "Obtainium/*" user-agent requests (over 10k over the past 24h) were allowed, with only several blocked or challenged.

The challenge reasons were:
- usage of tor exit nodes (which have extremely high malicious traffic rates and are therefore challenged or blocked)
- some urls are challenged to make certain bot access to them harder. I can explain a little more privately if you contact me on Telegram with the same username as I have here on Github.

Perhaps your test includes one of the challenged urls or you're using tor? Otherwise, please let me know on Telegram what you're querying for and perhaps your IP so I could look into it.

Furthermore, I can explain how APKUpdater/APKGrabber use our API and perhaps generate an API key for Obtainium so that you don't have to ping our feeds at all and instead batch request 100 apps at a time. In return, I ask only that Obtainium, just like APKGrabber and APKUpdater, continues to show users our download pages so that we can generate some ad revenue and pay for our server and operational costs.

**Comment by ImranR98** [comment]
Thanks @archon810, good to see that it's working for most people (I'm still seeing "Forbidden" with a couple of apps, "Garmin Connect IQ" and Netflix).
We're currently using the RSS feed, for example https://www.apkmirror.com/apk/netflix-inc/netflix/feed/.

> Furthermore, I can explain how APKUpdater/APKGrabber use our API and perhaps generate an API key for Obtainium so that you don't have to ping our feeds at all and instead batch request 100 apps at a time.

Yes, a proper API would probably be better than using the RSS feed. Obtainium currently only queries for one app at a time - so using the API would not be more efficient in that sense - but I'm guessing the API provides more info, is easier to parse, and may allow for other features like search.

Do you know how we could include the API key in Obtainium without making it public? Not sure how APKUpdater/APKGrabber do it. Or would it be okay to make that public? 

> I ask only that Obtainium, just like APKGrabber and APKUpdater, continues to show users our download pages so that we can generate some ad revenue and pay for our server and operational costs.

Yes, APKMirror is a "track-only" source and will stay that way.

**Comment by DwainZwerg** [comment]
So at least I still have forbidden errors too, all the time.

**Comment by FangPam** [comment]
> Latest version now correctly identifies Obtainium in the user agent (with no option to override). @archon810 please let us know if there's anything else that can be done. If not the source should be removed.


Hello, how to set the "Custom user-agent for APKMirror",I wasn't able to find the button.I'd be grateful if you could advise me.

**Comment by Noxba** [comment]
Are there any updates on this? 
APKMirror is still returning forbidden.

**Comment by DwainZwerg** [comment]
The only one who can do something about it is @archon810

**Comment by rakitanc** [comment]
Same case when adding the whatsapp from apkmirror
Forbidden error. Please fix it if yoy can. It would be really helpful

**Comment by Pocarai** [comment]
Same goes to me, I hope thos would be fixed Since Apkmirror is important

**Comment by archon810** [comment]
I tried to make sense of this app and how it uses apkmirror, but I just don't understand it.

With something like APK Updater or APK Grabber, it's straightforward - you install the app, open it up, it shows you a list of all of your installed apps, then you can choose APKmirror as one of the update sources, and it looks for updates to those apps on the site. Then you click on each update and it opens the relevant download page.

With Obtainium, there's no list of installed apps, so I don't really know what to do next to reproduce this issue.

**Comment by ImranR98** [comment]
@archon810 you go to the "add app" page in Obtainium and add the URL for any APKMirror app (for example the one in the issue description is https://www.apkmirror.com/apk/x-corp/twitter/).

**Comment by archon810** [comment]
OK, thanks. I was able to reproduce the Forbidden error, and I think it should be very easy to fix as it seems to be a bug in Obtainium.

Here's what I'm seeing in the logs when adding https://www.apkmirror.com/apk/x-corp/twitter:
- a request comes in for https://www.apkmirror.com/apk/x-corp/twitter/feed (without the trailing slash) with user agent "Obtainium/1.1.47". It's allowed, but it's a 301 redirect to https://www.apkmirror.com/apk/x-corp/twitter/feed/ (with the trailing slash).
- a second request for https://www.apkmirror.com/apk/x-corp/twitter/feed/ (with the trailing slash) comes in, but this one has a user agent "Dart/3.7 (dart:io)" which is blocked sitewide.

So, to fix this bug:
1. Make sure that any redirects are followed automatically, maintaining the user agent. Or at least that whatever follow-up requests are made, they keep the right user agent at all times and don't default to Dart.
2. To avoid 2 requests in the first place, append the trailing slash to the feed/ request right off the bat.

**Comment by ImranR98** [comment]
Done: https://github.com/ImranR98/Obtainium/releases/tag/v1.1.48
Thanks again @archon810

**Comment by archon810** [comment]
The fix worked for me as well.

**Comment by archon810** [comment]
This makes sense. The fix was to only append the / rather than properly handling redirects and handling the user agent.

**Comment by ImranR98** [comment]
https://github.com/ImranR98/Obtainium/releases/tag/v1.1.49
